listener.go

```diff
@@ -120,22 +136,36 @@ func (l *listener) onConn(conn net.Conn) {
 }
 
 func (l *listener) doOnConn(conn net.Conn) error {
-	start := time.Now()
-	// Read client init msg
-	initMsg := make([]byte, clientInitSize)
-	// Try to read start sequence
+	// We always wait for l.initMsgTimeout before responding to bad init messages. See
+	// listener.initMsgTimeout for more detail.
+
+	var (
+		start             = time.Now()
+		readDeadline      = start.Add(l.initMsgTimeout)
+		clearReadDeadline = func(c net.Conn) { c.SetReadDeadline(time.Time{}) }
+		initMsg           = make([]byte, clientInitSize)
+	)
+
+	conn.SetReadDeadline(readDeadline)
 	_, err := io.ReadFull(conn, initMsg)
 	if err != nil {
 		fullErr := log.Errorf("Unable to read client init msg %v after %v from %v ", err, time.Since(start), conn.RemoteAddr())
+		time.Sleep(time.Until(readDeadline))
+		clearReadDeadline(conn)
 		l.onError(conn, fullErr)
 		return fullErr
 	}
+
 	windowSize, maxPadding, cs, err := decodeClientInitMsg(l.serverPrivateKey, initMsg)
 	if err != nil {
 		fullErr := log.Errorf("Unable to decode client init msg from %v: %v", conn.RemoteAddr(), err)
+		// Continue reading from the peer until the read deadline.
+		ioutil.ReadAll(conn)
```

Comment:
Maybe `io.Copy(conn, ioutil.Discard)` is more efficient because it doesn't involve allocating space and storing data?

Reply:
Oh nice, I'll do that.  Thanks!

Reply:
Done